### PR TITLE
support array-like start/stop and add axis argument to linspace

### LINKED ISSUE #2446: Add support for array-like inputs for linspace
Currently, numpy supports array-like inputs while cupy only considers scalars for start & stop values of linspace function. It would be great to match the functionality. 

Sample code to reproduce the feature:

```
import numpy as np
a = np.linspace(0, 9, 10)
b = np.linspace(10, 19, 10)
result = np.linspace(a, b, 10)

#####################

import cupy as cp
a = cp.linspace(0, 9, 10)
b = cp.linspace(10, 19, 10)
result = cp.linspace(a, b, 10)

```

### PR DESCRIPTION
**Pull request by grlee77**
This supports array-like inputs and an `axis` argument in linspace as implemented in NumPy 1.16+.

closes #2446 

~I have marked this as WIP due to what is currently relatively slow performance for the array-like code path.~ (**edit:** slow for small `num`. For large enough `num`, cupy should always be faster)

A basic translation of the code from numpy 1.17 ends up being pretty slow for small arrays. It can be faster to just run `numpy.linspace` and transfer the result to the GPU in the case of array-like `start` and/or `stop` (as currently done in #2399).

For now, I check if both start & stop are scalars and run the current implementation as a fast path. This preserves performance for this common case, but results in some code duplication.

### CODE REVIEW

File: tests/cupy_tests/creation_tests/test_ranges.py [hunk 1]
```diff
@@ -145,6 +145,34 @@ def test_linspace_float_underflow(self, xp):
...
+    def test_linspace_array_start_stop(self, xp):
+        start = xp.array([-120, 120], dtype="int8")
+        stop = xp.array([100, -100], dtype="int8")
+        return xp.linspace(start, stop, num=50, dtype=float)
+
+    @testing.with_requires('numpy>=1.16')
+    @testing.numpy_cupy_array_equal()
+    def test_linspace_mixed_start_stop(self, xp):
+        start = 0.0
+        stop = xp.array([100, -100], dtype="int8")
```
Comment (asi1024) [inline tests/cupy_tests/creation_tests/test_ranges.py]: Could you parameterize dtype with `for_all_dtypes` or `for_all_dtypes_combination`? This test seems to fail if float32 array is given as the input.

Comment (grlee77) [reply tests/cupy_tests/creation_tests/test_ranges.py]: Okay, I have updated with more comprehensive testing.

There was indeed a subtle internal difference in the in dtypes being computed internally due to  #3035

If #3035 is addressed, then the refactoring done in 5dc9b9d should not be necessary.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
This CI failure looks unrelated to this PR. I will re-trigger after we fix the test configuration.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM! Thanks for your PR!

**Pull request merged**